dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,333 @@
+    //! \pre Driver is not initialized.
+    //! \return Operation status.
+    static err init();
+
+    //! Deinitialize serial driver and underlying platform bus.
+    //! \pre Driver is initialized.
+    //! \return Operation status.
+    static err deinit();
+
+    //! Obtains a byte from serial device.
+    //! \pre Driver is initialized.
+    //! \details Call will block if there is no data to return.
+    //! Otherwise data will be stored in the given argument and
+    //! method will return.
+    //! \param[out] byte Place for incoming byte. Will remain
+    //!                  unchanged if error occurs.
+    //! \return Operation status.
+    //! \retval err::nobufs Some data will be lost after the call completes.
+    //!
+    static err recv_byte(uint8_t &byte);
+
+    //! Gets buffer from serial device.
+    //! \pre Driver is initialized.
+    //! \details Call will block if there is no data to return.
+    //! Otherwise, buffer will be populated with avaliable data
+    //! and method will return.
+    //! \param[out]     buf Buffer to fill with data.
+    //! \param[in,out]  sz  Size of buffer. Will be updated with
+    //!                     amount of bytes written to the buffer.
+    //!                     In case of error, size will be updated
+    //!                     with bytes successfully stored in the buffer
+    //!                     before error occurred.
+    //! \return Operation status.
+    //! \retval err::nobufs Some data will be lost after the call completes.
+    static err recv_buf(uint8_t *buf, size_t &sz);
+
+    //! Sends byte to a serial device.
+    //! \pre Driver is initialized.
+    //! \details It may not block if internal buffering is applied.
+    //! It may also block if platform device is not yet ready to
+    //! send data.
+    //! \param[in] byte Byte to send.
+    //! \return Operation status.
+    static err send_byte(uint8_t byte);
+
+
+    //! Sends buffer to a serial stream.
+    //! \pre Driver is initialized.
+    //! \details It may not block if internal buffering is applied.
+    //! It may also block if platform device is not yet ready to
+    //! send data.
+    //! \param[in]      buf Buffer to send.
+    //! \param[in,out]  sz  Size of buffer. Will be updated with
+    //!                     amount of bytes sent In case of error,
+    //!                     size will be updated with amount of bytes
+    //!                     successfully sent before error occur.
+    //! \return Operation status.
+    //! \retval err:again The buffer was not consumed entirely.
+    static err send_buf(const uint8_t *buf, size_t &sz);
+
+private:
+    static void bus_handler(bus_channel ch, bus_event type, size_t total);
+
+    static bool m_is_inited;
+    static std::atomic_bool m_new_xfer_allowed;
+
+    static uint8_t m_rx_buffer[buf_size];
+    static std::atomic_size_t m_rx_write_iter; //!< Used for accepting data from the bus
```

Comment:
Consider this advice. Since you have 'invented' here smth like relaxed-lock queue, could you please add here a bit more info for future generations about domain, operation allowance and indicative conditions (predicates) that are associated with every iterator and flag?

For example, 
```
m_rx_write_iter 
constraints:
 - read only in normal (_not_ interrupt) execution mode, 
 - write/read in interrupt mode
 - m_rx_write_iter < buf_size is always true 
semantics:
 - denotes part of buffer that already populated with data
 - incremented with every byte reception in the event handler
indicative conditions:
 - in the normal execution mode m_rx_write_iter == m_rx_read_iter evaluates to true for case when buffer is empty, false otherwise 
 - in the interrupt execution mode, after byte reception and m_rx_write_iter increment m_rx_write_iter == m_rx_read_iter evaluates to true in case of overflow.
```

This is just rough example, pick your words and commenting approach. 

Still, it is just the advice, feel free to ignore it if you will consider it unnecessary (especially in the light that this info duplicates comments in event handler and recv/send methods)